### ISSUE TITLE
Equinox Decorators transforming scalar functions into arrays

### ISSUE THREAD

**Issue by cwoolfo1** [patrick-kidger/equinox#838]
I have been writing a custom loss function for a neural network that I am training. I have been using the equinox decorators in place of the jax jit decorator to just-in-time compile my loss function.

However, I have noticed that when I used the equinox decorators in place of the jax.jit, I get an entirely different result. My function will then return an array instead of scalar value.

Here is the code that caused this issue:

 ```
import jax
from jax import random
from jax import jit
import jax.numpy as jnp
import math
import equinox as eqx
# import external libraries

# @eqx.filter_jit
# @eqx.filter_grad
@jit
def loss_function(pred, x, dx, dy, dz):
    x_comp = (jnp.roll(x, shift=1, axis=0) + jnp.roll(x, shift=-1, axis=0) - 2*x)/(dx*dx)
    y_comp = (jnp.roll(x, shift=1, axis=1) + jnp.roll(x, shift=-1, axis=1) - 2*x)/(dy*dy)
    z_comp = (jnp.roll(x, shift=1, axis=2) + jnp.roll(x, shift=-1, axis=2) - 2*x)/(dz*dz)

    return jnp.mean( jnp.square( jnp.matmul( pred, jnp.squeeze(x_comp + y_comp + z_comp) ) - x ), axis=None )
```

**Comment by lockwo** [comment]
Can you provide a MVC? This example w/ jax.jit returns an array as well. 

```python
import jax
from jax import random
from jax import jit
import jax.numpy as jnp
import math
import equinox as eqx
# import external libraries

# @eqx.filter_jit
# @eqx.filter_grad
@jit
def loss_function(pred, x, dx, dy, dz):
   x_comp = (jnp.roll(x, shift=1, axis=0) + jnp.roll(x, shift=-1, axis=0) - 2*x)/(dx*dx)
   y_comp = (jnp.roll(x, shift=1, axis=1) + jnp.roll(x, shift=-1, axis=1) - 2*x)/(dy*dy)
   z_comp = (jnp.roll(x, shift=1, axis=2) + jnp.roll(x, shift=-1, axis=2) - 2*x)/(dz*dz)

   return jnp.mean( jnp.square( jnp.matmul( pred, jnp.squeeze(x_comp + y_comp + z_comp) ) - x ), axis=None )


loss_function(jnp.array([1.0, 1.0]), jnp.array([[[1.0, 2.0]]]), 0.1, 0.1, 0.1)
```

`Array(2.5, dtype=float32)`

**Comment by cwoolfo1** [comment]
This function takes in two, 3d arrays and returns a scalar value.

I have written a more complete test code below.

```
import jax.numpy as jnp
import jax
from jax import jit
import equinox as eqx

@eqx.filter_jit
@eqx.filter_grad
#@jit

def loss_function(pred, x, dx, dy, dz):
    """
Calculate the loss function for a given prediction.

Parameters:
- pred (ndarray): The prediction array.
- x (ndarray): The input array.
- dx (float): The step size in the x-direction.
- dy (float): The step size in the y-direction.
- dz (float): The step size in the z-direction.

Returns:
- float: The calculated loss.

"""
    x_comp = (jnp.roll(x, shift=1, axis=0) + jnp.roll(x, shift=-1, axis=0) - 2*x)/(dx*dx)
    y_comp = (jnp.roll(x, shift=1, axis=1) + jnp.roll(x, shift=-1, axis=1) - 2*x)/(dy*dy)
    z_comp = (jnp.roll(x, shift=1, axis=2) + jnp.roll(x, shift=-1, axis=2) - 2*x)/(dz*dz)
    Ax = jnp.squeeze(x_comp + y_comp + z_comp)
    # calculate the laplacian of the input
    return jnp.mean( jnp.square( jnp.matmul( pred, Ax ) - x ), axis=None )
    # A^-1 * Ax = x
    # calculate the x prediction and return the loss



dx = dy = dz = 1e-2
x = jnp.ones((50, 50, 50))
pred = 5*jnp.ones((50, 50, 50))

print(loss_function(pred, x, dx, dy, dz))
```

**Comment by lockwo** [comment]
That example, under jax.jit, returns an array:

```python
import jax.numpy as jnp
import jax
from jax import jit
import equinox as eqx

# @eqx.filter_jit
# @eqx.filter_grad
@jit
def loss_function(pred, x, dx, dy, dz):
    x_comp = (jnp.roll(x, shift=1, axis=0) + jnp.roll(x, shift=-1, axis=0) - 2*x)/(dx*dx)
    y_comp = (jnp.roll(x, shift=1, axis=1) + jnp.roll(x, shift=-1, axis=1) - 2*x)/(dy*dy)
    z_comp = (jnp.roll(x, shift=1, axis=2) + jnp.roll(x, shift=-1, axis=2) - 2*x)/(dz*dz)
    Ax = jnp.squeeze(x_comp + y_comp + z_comp)
    # calculate the laplacian of the input
    return jnp.mean( jnp.square( jnp.matmul( pred, Ax ) - x ), axis=None )
    # A^-1 * Ax = x
    # calculate the x prediction and return the loss



dx = dy = dz = 1e-2
x = jnp.ones((50, 50, 50))
pred = 5*jnp.ones((50, 50, 50))

print(type(loss_function(pred, x, dx, dy, dz)))
print(loss_function(pred, x, dx, dy, dz))
```

```
<class 'jaxlib.xla_extension.ArrayImpl'>
1.0
```

Equinox filter jit and jax jit return the same thing for me (both 1.0 scalar shape () array). I am running on CPU just in case that matters.

**Comment by cwoolfo1** [comment]
I placed this at the top of my code to use CPU:
```
jax.default_device(jax.devices("cpu")[0])

```

Are you using different library versions than me?
I have attached the output from running the above script along with my anaconda enviroment.
[enviroment.txt](https://github.com/user-attachments/files/16936024/enviroment.txt)
[log.txt](https://github.com/user-attachments/files/16936025/log.txt)

**Comment by lockwo** [comment]
I see your log is w/ the gradient estimation, is that intentional? The gradient will also be an array. 

This is my jax/eqx version: ('0.4.26', '0.11.5')

**Comment by cwoolfo1** [comment]
How did I not see that? Fixed!
Thank you.
